### ISSUE TITLE
angular 9 ivy input, OnChanges bug

### ISSUE THREAD

**Issue by luke-emmental** [angular/angular#37194]
# 🐞 bug report

### Affected Package
may be @angular/core

### Is this a regression?

angular 8 woking ok but angular 9 problem

#### expected  
only one input value change
#### issue
but other input value change

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-ivy-input-change-issue

## 🌍  Your Environment

stackbliz  dependencies

**Comment by pkozlowski-opensource** [comment]
tl;dr; my understanding is that the issue reported here is about arguments of the `ngOnChanges` callback.

@luke-emmental thnx for the reproduce scenario. I had a brief look at it and I must say that it is not 100% clear to me what is the observed vs. expected behaviour. Let me explain.

II've simplified a bit your scenario (mostly coping things to one file) here: https://stackblitz.com/edit/angular-ivy-input-change-issue-47nx1a?file=src%2Fapp%2Fapp.component.ts

I've just copied part of the scenario that you've marked as "this is not problem" but for me there is problem in this part as well :-) More precisely:
* expected: only `timeIn` is reported in `ngOnChanges`
* observed: both `objIn` (incorrect) and  `timeIn` (correct) is reported in `ngOnChanges`

There is definitively something going on here with `ngOnChanges` but I'm trying to be very precise about the scenario here. Could you guide me a bit more here on what is expected vs. observed?

**Comment by luke-emmental** [comment]
----
expected: only timeIn is reported in ngOnChanges
observed: both objIn (incorrect) and timeIn (correct) is reported in ngOnChanges
---- 
@pkozlowski-opensource  Yes! My intention is correct.

**Comment by luke-emmental** [comment]
I found `this is not problem` reason

in your case using `template`
I using `templateUrl`

The two cases give different results.

**Comment by destus90** [comment]
> The two cases give different results.

I don't see any difference in the observed behavior between the two reproductions above. The second one is a bit simpler.

**Comment by destus90** [comment]
Hmm, I have found an appropriate workaround there, just passing `ayncTimer` as `Object` without extracting a value.
https://stackblitz.com/edit/angular-ivy-input-change-issue-ztz4kh?file=src/app/app.component.ts

**Comment by luke-emmental** [comment]
> I found `this is not problem` reason
> 
> in your case using `template`
> my case using `templateUrl`
> 
> The two cases give different results.

 umm... sorry.  It was wrong

**Comment by pkozlowski-opensource** [comment]
@destus90 ha! I think that you are onto something! I've just observed that just removing `?` also makes things work as expected. That is, replacing `[timeIn]="(ayncTimer|async)?.t` with `[timeIn]="(ayncTimer|async).t` <- notice lack of ?) gets us to the expected value in the `OnChanges` callback (it will error initially as the first emitted value is `null` but then it starts to work OK: https://stackblitz.com/edit/angular-ivy-input-change-issue-vnypga?file=src/app/app.component.ts

To sum up, `ngOnChanges` is over-reporting changes under some circumstances. In the reproduction scenario it means that:
* `[timeIn]="(ayncTimer|async)?.t` - is WRONG (expected notification with `timeIn`, actual notification with `timeIn` and `objIn`)
* `[timeIn]="(ayncTimer|async).t` is OK (expected notification with `timeIn` only)


It is not clear to me if this is a compiler bug (we are generating different code for the 2 above cases) or the runtime bug. This will need more root-cause analysis but at this point it looks to me like a confirmed bug in ivy.

**Comment by kara** [comment]
OK, I think I've gotten to the bottom of this one. It turns out that when you use a `?` operator, the compiler is pulling out the value of the `timeIn` binding like this (into `currVal_1`):

```js
const currVal_1 = (tmp_1_0 = i0.ɵɵpipeBind1(1, 2, ctx.ayncTimer)) == null ? null : tmp_1_0.t;
i0.ɵɵproperty("objIn", ctx.objValue)("timeIn", currVal_1);
```

This is a problem because the pipe inside `timeIn` is now executed out of order - before the `objValue` binding that comes before it. This matters because the pipe binding code gets the current binding index value, but it will now point to `objValue` because we haven't yet executed that instruction. As such, the pipe will invalidate the `objValue` binding instead of the `timeIn` binding and cause a property update, which then triggers the `ngOnChanges` machinery (which is actually working as expected).

**Comment by mistic100** [comment]
Another example of it helps, using property setters

https://stackblitz.com/edit/angular-ivy-xkcbs1

In the setter of "value1" I increment a counter, this counter is increment when "value2" changes.

**Comment by YuJM** [comment]
umm.. angular 9.1.11
This issue is still valid.

**Comment by JoostK** [comment]
While working on a fix for this issue, I noticed that it only occurs for pipes that return `WrappedValue`. Given that the `async` pipe in Angular 10 is no longer using `WrappedVale`, this issue has resolved itself for the `async` pipe.

Opened PR #37911 to address the issue in general.
